R/drive_path.R

```diff
@@ -1,91 +1,193 @@
-  ## all pieces where that is known
-  hits <- drive_list(
-    fields = "files/parents,files/name,files/mimeType,files/id",
-    q = form_query(path_pieces, leaf_is_folder = grepl("/$", path)),
-    verbose = FALSE
-  )
+  if (is.null(.rships)) {
+    ## query restricts to names in path_pieces and, for all pieces that are
+    ## known to be folder, to mimeType = folder
+    .rships <- drive_list(
+      fields = "files/parents,files/name,files/mimeType,files/id",
+      q = form_query(path_pieces, leaf_is_folder = grepl("/$", path)),
+      verbose = FALSE
+    )
+    ## fetch fileId of user's My Drive root folder
+    .root <- root_id()
+  }
 
-  if (!all(path_pieces %in% hits$name)) {
-    stop(glue::glue("The path '{path}' does not exist.", call. = FALSE))
+  ## revise target path to be longest partial path that is compatible
+  ## with folder names that exist
+  nm_detected <- purrr::set_names(path_pieces %in% .rships$name, path_pieces)
+  d <- last_all(nm_detected)
+  if (d == 0 || (!partial_ok && d < length(path_pieces))) {
+    return(null_path())
   }
-  ## guarantee: we have found at least one item with correct name for
-  ## each piece of path
 
-  ## leaf candidate(s)
-  leaf_id <- hits$id[hits$name == last(path_pieces)]
+  repeat {
+    ## leaf candidate(s)
+    leaf_id <- .rships$id[.rships$name == path_pieces[d]]
 
-  ## for each candidate, enumerate all upward paths, hopefully to root
-  root_path <- leaf_id %>%
-    purrr::map(pth, kids = hits$id, elders = hits$parents, stop_value = root_id)
+    ## for each candidate, enumerate all rooted paths
+    leaf_tbl <- leaf_id %>%
+      purrr::map(pth_tbl, .rships = .rships, stop_value = .root)
+    leaf_tbl <- do.call(rbind, leaf_tbl)
 
-  ## put into a tibble, one row per candidate path
-  leaf_tbl <- tibble::tibble(
-    id = rep(leaf_id, lengths(root_path)),
-    root_path = purrr::flatten(root_path)
-  )
-  leaf_tbl$path <- purrr::map_chr(
-    leaf_tbl$root_path,
-    ~ glue::collapse(rev(hits$name[match(crop(.x, 1), hits$id)]), sep = "/")
-  )
+    ## require path to match target, in manner appropriate to partial_ok
+    ## glue::glue() gives me this: length 0 in --> length 0 out
+    path_matches <- purrr::map_lgl(
+      glue::glue("^{leaf_tbl$path}{if (partial_ok) '' else '$'}"),
+      grepl, x = strip_slash(path)
+    )
+    leaf_tbl <- leaf_tbl[path_matches, ]
 
-  ## retain candidate paths that match input path
-  keep_me <- which(strip_slash(path) == leaf_tbl$path)
+    if (d == 1 || !partial_ok || nrow(leaf_tbl) > 0) {
+      break
+    }
+    d <- d - 1
+  }
 
-  if (length(keep_me) > 1) {
-    line0 <- glue::glue("The path '{path}' identifies more than one file:")
-    lines <- glue::glue_data(
-      hits[hits$id %in% leaf_id[keep_me], ],
-      "File of type {type}, with id {id}."
-    )
+  ## ensure path_orig gets added, even if there are zero rows
+  leaf_tbl$path_orig <- rep_len(path, nrow(leaf_tbl))
+  leaf_tbl
```

Comment:
Once we revisit the object that holds file metadata (maybe becoming a dribble?), that should probably be what this returns. `drive_path()` is an exported function that exposes this function and it should probably drop the `path_orig` variable. I figure we can refine this when we create the dribble class and methods.